Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,480 @@
+
+            // We can choose whichever alternative we want. In general, we should choose the last alternative that know we can display.
+            // RFC 2046 Section 5.1.4 https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.4
+            auto chosen_alternative = alternatives.last_matching([this](auto& alternative) {
+                return is_supported_alternative(alternative);
+            });
+
+            if (alternatives.is_empty()) {
+                GUI::MessageBox::show(window(), String::formatted("Displaying this type of e-mail is currently unsupported."), "Unsupported", GUI::MessageBox::Type::Information);
+                return;
+            }
+            selected_alternative_position = chosen_alternative.value().position;
+            selected_alternative_encoding = chosen_alternative.value().body_structure.encoding;
+        });
+
+    if (selected_alternative_position.is_empty()) {
+        // An error occurred above, return.
+        return;
+    }
+
+    fetch_command = IMAP::FetchCommand {
+        .sequence_set { { id_of_email_to_load, id_of_email_to_load } },
+        .data_items = {
+            IMAP::FetchCommand::DataItem {
+                .type = IMAP::FetchCommand::DataItemType::BodySection,
+                .section = IMAP::FetchCommand::DataItem::Section {
+                    .type = IMAP::FetchCommand::DataItem::SectionType::Parts,
+                    .parts = selected_alternative_position,
+                },
+                .partial_fetch = false,
+            },
+        },
+    };
+
+    fetch_response = m_imap_client->fetch(fetch_command, false)->await().release_value();
+
+    if (fetch_response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to retrieve the body of the selected e-mail. The server says: '{}'", fetch_response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve the selected e-mail. The server says: '{}'", fetch_response.response_text()));
+        return;
+    }
+
+    auto& encoded_data = fetch_response.data()
+                             .fetch_data()
+                             .last()
+                             .get<IMAP::FetchResponseData>()
+                             .body_data()
+                             .find_if([](Tuple<IMAP::FetchCommand::DataItem, Optional<String>>& data) {
+                                 const auto data_item = data.get<0>();
+                                 return data_item.section.has_value() && data_item.section->type == IMAP::FetchCommand::DataItem::SectionType::Parts;
+                             })
+                             ->get<1>()
+                             .value();
+
+    String decoded_data;
+
+    // FIXME: String uses char internally, so 8bit shouldn't be stored in it.
+    //        However, it works for now.
+    if (selected_alternative_encoding.equals_ignoring_case("7bit") || selected_alternative_encoding.equals_ignoring_case("8bit")) {
```

Comment:
Without knowing any of the IMAP spec, I assume "8bit" is one of the legacy 8-bit encodings (such as "windows-1252"). If this is the case, you could/should use the relevant decoder from LibTextCodec to decode the String.

Reply:
8-bit is the same as 7-bit, other than it allows up to 255 instead of just 127. https://datatracker.ietf.org/doc/html/rfc2045#section-2.8
The charset is specified separately.